google/cloud/aiplatform/models.py

```diff
@@ -215,10 +244,548 @@ def upload(
+                -  ``n1-standard-32``
+                -  ``n1-highmem-2``
+                -  ``n1-highmem-4``
+                -  ``n1-highmem-8``
+                -  ``n1-highmem-16``
+                -  ``n1-highmem-32``
+                -  ``n1-highcpu-2``
+                -  ``n1-highcpu-4``
+                -  ``n1-highcpu-8``
+                -  ``n1-highcpu-16``
+                -  ``n1-highcpu-32``
+                Not specifying machine type will result in model to be deployed
+                with automatic resources.
+            min_replica_count (int):
+                Optional. The minimum number of machine replicas this deployed
+                model will be always deployed on. If traffic against it increases,
+                it may dynamically be deployed onto more replicas, and as traffic
+                decreases, some of these extra replicas may be freed.
+            max_replica_count (int):
+                Optional. The maximum number of replicas this deployed model may
+                be deployed on when the traffic against it increases. If requested
+                value is too large, the deployment will error, but if deployment
+                succeeds then the ability to scale the model to that many replicas
+                is guaranteed (barring service outages). If traffic against the
+                deployed model increases beyond what its replicas at maximum may
+                handle, a portion of the traffic will be dropped. If this value
+                is not provided, the smaller value of min_replica_count or 1 will
+                be used.
+            metadata (Sequence[Tuple[str, str]]):
+                Optional. Strings which should be sent along with the request as
+                metadata.
+            timeout (int):
+                Optional. How long (in seconds) to wait for the endpoint creation
+                operation to complete. If None, wait indefinitely.
+        Returns:
+            endpoint ("Endpoint"):
+                Endpoint with the deployed model.
+
+        """
+        if endpoint_obj is None:
+            display_name = self._gca_resource.display_name + "_endpoint"
+            endpoint_obj = Endpoint.create(display_name=display_name)
+
+            # TODO(b/171631203) queue deploy instead of block
+            @retry.Retry(deadline=timeout)
+            def endpoint_exist():
+                if endpoint_obj._gca_resource is None:
+                    raise Exception("Endpoint not yet created.")
+
+            endpoint_exist()
+
+        endpoint_obj.deploy(
+            self,
+            deployed_model_display_name=deployed_model_display_name,
+            traffic_percentage=traffic_percentage,
+            machine_type=machine_type,
+            min_replica_count=min_replica_count,
+            max_replica_count=max_replica_count,
+            metadata=metadata,
+        )
+
+        return endpoint_obj
+
+
+class Endpoint(base.AiPlatformResourceNoun):
```

Comment:
@sasha-gitg I recall you wanted endpoint to be in the model file. does this PR change your mind?

Reply:
The argument to keep these in the same file was to avoid a circular imports because Model.deploy takes Endpoint and Endpoint.deploy takes Model and there is an ask to take their respective resource names and instantiate the classes in the methods.

Reply:
Are these two mutually imported only in typing annotations? (in which case we might be able to still split them into separate files)

Reply:
I agree if it was just type annotations but there was an ask during review to support resource names so we would need to instantiate the classes within the methods. ie:

```python

class Model:
    def deploy(endpoint: Union[Endpoint, str], ....):
        if isinstance(endpoint, str):
            endpoint = Endpoint(endpoint)
        ...
```

This isn't P0 but we should take it into account to avoid a future refactor.

Reply:
we'll table this for the future